--- php-5.3.1/configure.in	2009-11-18 21:43:33.000000000 +0200
+++ php-5.3.1/configure.in.mod	2010-02-02 10:13:09.000000000 +0200
@@ -426,7 +426,6 @@
 arpa/inet.h \
 arpa/nameser_compat.h \
 arpa/nameser.h \
-assert.h \
 crypt.h \
 dns.h \
 fcntl.h \
@@ -468,7 +467,8 @@
 sys/utsname.h \
 sys/ipc.h \
 dlfcn.h \
-assert.h
+assert.h \
+crt_externs.h \
 ],[],[],[
 #ifdef HAVE_SYS_PARAM_H
 #include <sys/param.h>
--- php-5.3.1/ext/spl/spl_dllist.h	2008-12-31 13:15:49.000000000 +0200
+++ php-5.3.1/ext/spl/spl_dllist.h.mod	2010-02-02 10:15:22.000000000 +0200
@@ -24,9 +24,10 @@
 #include "php.h"
 #include "php_spl.h"
 
-PHPAPI zend_class_entry *spl_ce_SplDoublyLinkedList;
-PHPAPI zend_class_entry *spl_ce_SplQueue;
-PHPAPI zend_class_entry *spl_ce_SplStack;
+extern PHPAPI zend_class_entry *spl_ce_SplDoublyLinkedList;
+extern PHPAPI zend_class_entry *spl_ce_SplQueue;
+extern PHPAPI zend_class_entry *spl_ce_SplStack;
+
 
 PHP_MINIT_FUNCTION(spl_dllist);
 
--- php-5.3.1/ext/spl/spl_fixedarray.h	2008-12-31 13:15:49.000000000 +0200
+++ php-5.3.1/ext/spl/spl_fixedarray.h.mod	2010-02-02 10:16:26.000000000 +0200
@@ -22,7 +22,7 @@
 #ifndef SPL_FIXEDARRAY_H
 #define SPL_FIXEDARRAY_H
 
-PHPAPI zend_class_entry *spl_ce_SplFixedArray;
+extern PHPAPI zend_class_entry *spl_ce_SplFixedArray;
 
 PHP_MINIT_FUNCTION(spl_fixedarray);
 
--- php-5.3.1/ext/spl/spl_heap.h	2008-12-31 13:15:49.000000000 +0200
+++ php-5.3.1/ext/spl/spl_heap.h.mod	2010-02-02 10:17:34.000000000 +0200
@@ -24,11 +24,10 @@
 #include "php.h"
 #include "php_spl.h"
 
-PHPAPI zend_class_entry *spl_ce_SplHeap;
-PHPAPI zend_class_entry *spl_ce_SplMinHeap;
-PHPAPI zend_class_entry *spl_ce_SplMaxHeap;
-
-PHPAPI zend_class_entry *spl_ce_SplPriorityQueue;
+extern PHPAPI zend_class_entry *spl_ce_SplHeap;
+extern PHPAPI zend_class_entry *spl_ce_SplMinHeap;
+extern PHPAPI zend_class_entry *spl_ce_SplMaxHeap;
+extern PHPAPI zend_class_entry *spl_ce_SplPriorityQueue;
 
 PHP_MINIT_FUNCTION(spl_heap);
 
--- php-5.3.1/main/php.h	2009-06-26 18:44:19.000000000 +0300
+++ php-5.3.1/main/php.h.mod	2010-02-02 10:20:57.000000000 +0200
@@ -263,7 +263,13 @@
 #if !defined(PHP_WIN32)
 #define PHP_SLEEP_NON_VOID
 #define php_sleep sleep
+#if HAVE_CRT_EXTERNS_H
+#include <crt_externs.h>
+#define environ (*_NSGetEnviron())
+#else
 extern char **environ;
+#endif
+
 #endif	/* !defined(PHP_WIN32) */
 
 #ifdef PHP_PWRITE_64
--- php-5.3.1/sapi/cli/config.m4	2008-09-01 16:15:31.000000000 +0300
+++ php-5.3.1/sapi/cli/config.m4.mod	2010-02-02 10:27:35.000000000 +0200
@@ -20,9 +20,6 @@
       BUILD_CLI="echo '\#! .' > php.sym && echo >>php.sym && nm -BCpg \`echo \$(PHP_GLOBAL_OBJS) \$(PHP_CLI_OBJS) | sed 's/\([A-Za-z0-9_]*\)\.lo/\1.o/g'\` | \$(AWK) '{ if (((\$\$2 == \"T\") || (\$\$2 == \"D\") || (\$\$2 == \"B\")) && (substr(\$\$3,1,1) != \".\")) { print \$\$3 } }' | sort -u >> php.sym && \$(LIBTOOL) --mode=link \$(CC) -export-dynamic \$(CFLAGS_CLEAN) \$(EXTRA_CFLAGS) \$(EXTRA_LDFLAGS_PROGRAM) \$(LDFLAGS) -Wl,-brtl -Wl,-bE:php.sym \$(PHP_RPATHS) \$(PHP_GLOBAL_OBJS) \$(PHP_CLI_OBJS) \$(EXTRA_LIBS) \$(ZEND_EXTRA_LIBS) -o \$(SAPI_CLI_PATH)"
     fi
     ;;
-  *darwin*)
-    BUILD_CLI="\$(CC) \$(CFLAGS_CLEAN) \$(EXTRA_CFLAGS) \$(EXTRA_LDFLAGS_PROGRAM) \$(LDFLAGS) \$(NATIVE_RPATHS) \$(PHP_GLOBAL_OBJS:.lo=.o) \$(PHP_CLI_OBJS:.lo=.o) \$(PHP_FRAMEWORKS) \$(EXTRA_LIBS) \$(ZEND_EXTRA_LIBS) -o \$(SAPI_CLI_PATH)"
-    ;;
   *netware*)
     BUILD_CLI="\$(LIBTOOL) --mode=link \$(CC) -export-dynamic \$(CFLAGS_CLEAN) \$(EXTRA_CFLAGS) \$(EXTRA_LDFLAGS_PROGRAM) \$(LDFLAGS) \$(PHP_RPATHS) \$(PHP_CLI_OBJS) \$(EXTRA_LIBS) \$(ZEND_EXTRA_LIBS) -Lnetware -lphp5lib -o \$(SAPI_CLI_PATH)"
     ;;
--- php-5.3.1/main/php_getopt.h	2008-12-31 13:15:49.000000000 +0200
+++ php-5.3.1/main/php_getopt.h.mod	2010-02-02 10:22:52.000000000 +0200
@@ -41,7 +41,7 @@
 
 BEGIN_EXTERN_C()
 /* holds the index of the latest fetched element from the opts array */
-PHPAPI int php_optidx;
+extern PHPAPI int php_optidx;
 PHPAPI int php_getopt(int argc, char* const *argv, const opt_struct opts[], char **optarg, int *optind, int show_err, int arg_start);
 END_EXTERN_C()
 
