Index: php-src/ext/json/utf8_to_utf16.c
===================================================================
--- php-src/ext/json/utf8_to_utf16.c:1.1	Tue Jan 31 08:59:06 2006
+++ php-src/ext/json/utf8_to_utf16.c	Thu Aug 21 18:10:48 2008
@@ -40,7 +40,7 @@
     for (;;) {
         c = utf8_decode_next(&utf8);
         if (c < 0) {
-            return UTF8_END ? the_index : UTF8_ERROR;
+            return (c == UTF8_END) ? the_index : UTF8_ERROR;
         }
         if (c < 0x10000) {
             w[the_index] = (unsigned short)c;
