--- php-5.2.5/sapi/cgi/cgi_main.c	2007-11-01 17:23:14.000000000 +0200
+++ php-5.2.5-patched/sapi/cgi/cgi_main.c	2008-05-01 13:30:44.000000000 +0300
@@ -232,6 +232,7 @@
 	zend_llist sorted_exts;
 
 	zend_llist_copy(&sorted_exts, &zend_extensions);
+	sorted_exts.dtor = NULL;
 	zend_llist_sort(&sorted_exts, extension_name_cmp TSRMLS_CC);
 	zend_llist_apply_with_argument(&sorted_exts, (llist_apply_with_arg_func_t) print_extension_info, NULL TSRMLS_CC);
 	zend_llist_destroy(&sorted_exts);
