--- php-5.2.9/Zend/zend_execute_API.c.orig	2009-05-17 08:54:03.207490832 +0300
+++ php-5.2.9/Zend/zend_execute_API.c	2009-05-17 09:05:49.818753808 +0300
@@ -1475,7 +1475,7 @@
 		t_r.it_value.tv_sec = seconds;
 		t_r.it_value.tv_usec = t_r.it_interval.tv_sec = t_r.it_interval.tv_usec = 0;
 
-#	ifdef __CYGWIN__
+#	if defined(__CYGWIN__) || defined(AIX_PASE)
 		setitimer(ITIMER_REAL, &t_r, NULL);
 		signal(SIGALRM, zend_timeout);
 		sigemptyset(&sigset);
@@ -1506,7 +1506,7 @@
 
 		no_timeout.it_value.tv_sec = no_timeout.it_value.tv_usec = no_timeout.it_interval.tv_sec = no_timeout.it_interval.tv_usec = 0;
 
-#ifdef __CYGWIN__
+#if defined(__CYGWIN__) || defined(AIX_PASE)
 		setitimer(ITIMER_REAL, &no_timeout, NULL);
 #else
 		setitimer(ITIMER_PROF, &no_timeout, NULL);
